Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs

```diff
@@ -19,31 +19,73 @@ public class WorkerConnection : IDisposable
         private readonly ConcurrentDictionary<long, TaskHandler> requestsToComplete = new ConcurrentDictionary<long, TaskHandler>();
         private Connection connection;
         private Task metricsTask;
-        private CancellationTokenSource metricsTcs = new CancellationTokenSource();
-        private string workerId;
+        private CancellationTokenSource metricsCts = new CancellationTokenSource();
         private readonly object connectionLock = new object();
+        private readonly CancellationTokenSource processOpsCts = new CancellationTokenSource();
+        private readonly BlockingCollection<OpList> ops = new BlockingCollection<OpList>();
+        private Task processOpsTask;
 
         private WorkerConnection(Connection connection)
         {
             this.connection = connection ?? throw new ArgumentNullException(nameof(connection));
+            WorkerId = connection.GetWorkerId();
+
+            StartBackgroundOpsGathering();
         }
 
-        public string WorkerId
+        private void StartBackgroundOpsGathering()
         {
-            get
+            processOpsTask = Task.Factory.StartNew(() =>
             {
-                if (string.IsNullOrEmpty(workerId))
+                // We have to resort to polling here since connection.GetOpList doesn't provide a means of cancellation
+                try
                 {
-                    // ReSharper disable once InconsistentlySynchronizedField
-                    workerId = connection.GetWorkerId();
-                }
+                    while (!processOpsCts.Token.IsCancellationRequested)
+                    {
+                        if (GetConnectionStatusCode() != ConnectionStatusCode.Success)
+                        {
+                            break;
+                        }
 
-                return workerId;
-            }
+                        OpList opList;
+                        lock (connectionLock)
+                        {
+                            var rawOps = connection.GetOpList(1);
+                            if (rawOps.GetOpCount() == 0)
+                            {
+                                continue;
+                            }
+
+                            opList = new OpList(rawOps);
+                        }
+
+                        // NB: finish processing the OpList before adding it to the queue, since it will be Disposed on whatever thread is enumerating the ops collection.
+                        CompleteCommands(opList);
+                        ops.Add(opList);
+                    }
+                }
+                finally
+                {
+                    ops.CompleteAdding();
+                }
+            }, TaskCreationOptions.LongRunning);
         }
 
+        public string WorkerId { get; }
+
         public void Dispose()
```

Comment:
Should this be calling `CancelCommands` as well?

Reply:
It does, lower in the function

Reply:
Oh I'm just blind (or Github collapsed the diff)